### DISCUSSION TITLE
about js interoperability

### DISCUSSION THREAD

**Discussion by rraammiinn** [vlang/v#17613] [Questions and Answers]
can vls interact with html dom and could be there a solution to use js libraries in a v program when you wanna compile it to js ?

**Comment by JalonSolov** [comment]
Why would the V language server need to interact with the HTML DOM?

**Reply by rraammiinn** [reply]
so it can be used for web development .
typescript added som features to js but didn't remove it's bad features  (for example promise syntax for me) .

**Reply by JalonSolov** [reply]
I still don't understand.  The language server is simply that - a server.  It doesn't interact with the screen at all.  That is the job of the editor calling the server.

So if you did a web editor that talked to the language server, it should work fine, with no changes needed to the server.

**Reply by rraammiinn** [reply]
I'm talking about autocompletion so it can detect dom elements .
I thought this responsibility has been moved from editor plugins to the language services .

**Reply by JalonSolov** [reply]
The editor would query the language server to find the info, then the _editor_ will display it.

The server still doesn't do any displaying.  It just provides the info.

**Reply by rraammiinn** [reply]
if there be a v library for dom manipulation how it can find out what elements are there in the targeted document ?

**Reply by JalonSolov** [reply]
How many times must I repeat myself?  The editor sends a query.  Something like "What language features start with 'os.'?"

vls does NOT look at the document directly.  The editor must send it to the server.

The server maintains a copy of the _text_ of the document.  It doesn't have any idea about how that text will be display, be it screen, dom, or chalkboard.

Every change a user makes is sent to the server, which updates it's internal copy of the document.

<sigh>  You know what?  I don't want to keep trying to explain this when there are lots of explanations on the web.  Just search.  I'll even help with a place to start:

https://microsoft.github.io/language-server-protocol/overviews/lsp/overview/

**Comment by rraammiinn** [comment]
if some day wasm can manipulate dom directly it could be even better .
